src/solvers/dg/interpolation.jl

```diff
@@ -1,88 +1,220 @@
+  # 2D
+  # optimized version of multiply_dimensionwise_naive
+  size_out = size(matrix, 1)
+  n_vars   = size(data_in, 1)
+  data_out = zeros(promote_type(eltype(data_in), eltype(matrix)), n_vars, size_out, size_out)
 
-  for k in 1:n_nodes_out, j in 1:n_nodes_out, i in 1:n_nodes_out
-    for v in 1:n_vars
-      acc = zero(eltype(data_out))
-      for kk in 1:n_nodes_in, jj in 1:n_nodes_in, ii in 1:n_nodes_in
-        acc += vandermonde[i, ii] * vandermonde[j, jj] * vandermonde[k, kk] * data_in[v, ii, jj, kk]
-      end
-      data_out[v, i, j, k] = acc
-    end
-  end
+  multiply_dimensionwise!(data_out, matrix, data_in)
 
   return data_out
 end
 
+function multiply_dimensionwise(matrix::AbstractMatrix, data_in::AbstractArray{<:Any, 4})
+  # 3D
+  # optimized version of multiply_dimensionwise_naive
+  size_out = size(matrix, 1)
+  n_vars   = size(data_in, 1)
+  data_out = zeros(promote_type(eltype(data_in), eltype(matrix)), n_vars, size_out, size_out, size_out)
+
+  multiply_dimensionwise!(data_out, matrix, data_in)
+
+  return data_out
+end
+
+
+# in-place version of multiply_dimensionwise
+function multiply_dimensionwise!(data_out::AbstractArray{<:Any, 2}, matrix::AbstractMatrix,
+                                 data_in ::AbstractArray{<:Any, 2})
+  @tullio threads=false data_out[v, i] = matrix[i, ii] * data_in[v, ii]
+
+  return nothing
+end
+
+function multiply_dimensionwise!(data_out::AbstractArray{<:Any, 2}, matrix1::AbstractMatrix,
+                                 data_in1::AbstractArray{<:Any, 2}, matrix2::AbstractMatrix,
+                                 data_in2::AbstractArray{<:Any, 2})
+  @tullio threads=false data_out[v, i] = matrix1[i, ii] * data_in1[v, ii] + matrix2[i, ii] * data_in2[v, ii]
+
+  return nothing
+end
+
+function multiply_dimensionwise!(data_out::AbstractArray{<:Any, 3}, matrix::AbstractMatrix,
+                                 data_in:: AbstractArray{<:Any, 3},
+                                 tmp1=zeros(eltype(data_out), size(data_out, 1), size(matrix, 1), size(matrix, 2)))
+
+  # Interpolate in x-direction
+  @tullio threads=false tmp1[v, i, j]     = matrix[i, ii] * data_in[v, ii, j]
+
+  # Interpolate in y-direction
+  @tullio threads=false data_out[v, i, j] = matrix[j, jj] * tmp1[v, i, jj]
+
+  return nothing
+end
+
+function multiply_dimensionwise!(data_out::AbstractArray{<:Any, 3},
+                                 matrix1::AbstractMatrix, matrix2::AbstractMatrix,
+                                 data_in:: AbstractArray{<:Any, 3},
+                                 tmp1=zeros(eltype(data_out), size(data_out, 1), size(matrix1, 1), size(matrix1, 2)))
+
+  # Interpolate in x-direction
+  @tullio threads=false tmp1[v, i, j]     = matrix1[i, ii] * data_in[v, ii, j]
+
+  # Interpolate in y-direction
+  @tullio threads=false data_out[v, i, j] = matrix2[j, jj] * tmp1[v, i, jj]
+
+  return nothing
+end
+
+function add_multiply_dimensionwise!(data_out::AbstractArray{<:Any, 3},
```

Comment:
There should be at least one docstring for `add_multiply_dimensionwise!` or `add_multiply_dimensionwise` (maybe there is somewhere and I just missed it)